slimta/edge/wsgi.py

```diff
@@ -89,13 +90,25 @@ class WsgiResponse(Exception):
 
     """
 
-    def __init__(self, status, headers=None, data=None):
+    def __init__(self, status, headers=[], data=[]):
         super(WsgiResponse, self).__init__(status)
         self.status = status
-        self.headers = headers or []
+        self.headers = [(_unicode_to_bytes(k), _unicode_to_bytes(v))
+                        for k, v in headers]
         self.data = data or []
 
 
+def _unicode_to_bytes(s):
+    """Encodes headers to iso-8859-1 bytes
+
+    https://www.python.org/dev/peps/pep-0333/#unicode-issues
+    """
+    assert isinstance(s, six.string_types), (
+        "{} is supposed to be an unicode string, not {}".format(type(s))
+    )
```

Comment:
With py2, we can have a py2 `str` arg, which is not unicode, and fail at *encoding* stage (UnicodeDecodeError) where we may expect either *AttributeError*. I agree this is not embracing  duck-typing style, but I consider we are in a case (bytes vs str in py2), where too much magic occurs and that making things a bit more strict with checking code helps development and troubleshooting. 

It's a global design/style issue, and I will let you the final word anyway, what do you think ?